File: tests/ruby/src/T016_ResourceGenerator.rb
#!/usr/bin/env ruby

require 'libgmsec_ruby'

load 'Test.rb'


class Test_ResourceGenerator < Test

    def initialize(args)
        super(args)

        @@stdFields = get_standard_fields()
    end


    def test()
        test_create_1
        test_create_2
        test_start
        test_stop
        test_set_field
    end


    def test_create_1()
        Libgmsec_ruby::Log::info("Test create 1")

        # Nominal test
        rsrcgen = Libgmsec_ruby::ResourceGenerator::create(get_config(), 5, 1, 10)
        check("ResourceGenerator should not be running", rsrcgen.is_running() == false)
        Libgmsec_ruby::ResourceGenerator::destroy(rsrcgen)

        # Off-nominal test(s)
        begin
            config = Libgmsec_ruby::Config::new(get_config())
            config.add_value("mw-id", "bogus-mw")
            Libgmsec_ruby::ResourceGenerator::create(config, 5, 1, 10)
            check("An exception was expected", false)
        rescue GmsecException => e
            check(e.message, e.message.include?("Unable to load"))
        end
    end


    def test_create_2()
        Libgmsec_ruby::Log::info("Test create 2")

        emptyFieldList = Libgmsec_ruby::FieldArray.new()

        # Nominal test (empty, populated, and null list of fields)
        rsrcgen1 = Libgmsec_ruby::ResourceGenerator::create(get_config(), 5, 1, 10, emptyFieldList)
        check("ResourceGenerator should not be running", rsrcgen1.is_running() == false)
        Libgmsec_ruby::ResourceGenerator::destroy(rsrcgen1)

        rsrcgen2 = Libgmsec_ruby::ResourceGenerator::create(get_config(), 5, 1, 10, @@stdFields)
        check("ResourceGenerator should not be running", rsrcgen2.is_running() == false)
        Libgmsec_ruby::ResourceGenerator::destroy(rsrcgen2)

        # Off-nominal test(s)
        begin
            config = Libgmsec_ruby::Config::new(get_config())
            config.add_value("mw-id", "bogus-mw")
            Libgmsec_ruby::ResourceGenerator::create(config, 5, 1, 10, @@stdFields)
            check("An exception was expected", false)
        rescue GmsecException => e
            check(e.message, e.message.include?("Unable to load"))
        end
    end


    def test_start()
        Libgmsec_ruby::Log::info("Test start()")

        config = Libgmsec_ruby::Config::new(get_config())
        pubRate = 1

        rsrcgen = Libgmsec_ruby::ResourceGenerator::create( config, pubRate, 1, 10, @@stdFields )

        # To ensure this resource message is unique, we set the COMPONENT field
        rsrcgen.set_field( Libgmsec_ruby::StringField.new("COMPONENT", get_unique_component()) )

        rsrcgen.start()

        check("ResourceGenerator should be running", rsrcgen.is_running())

        verify_resource_message(config, pubRate)

        rsrcgen.stop()

        # Allow time for the RSRC-gen thread to stop
        sleep(4)

        Libgmsec_ruby::ResourceGenerator::destroy(rsrcgen)

        # Off-nominal tests
        Libgmsec_ruby::Log::info("Off-nominal cases...")
        config.add_value("gmsec-msg-content-validate", "true")

        rsrcgen2 = Libgmsec_ruby::ResourceGenerator::create(config, 1, 1, 10, get_standard_fields)

        # Add bogus field using a Field
        begin
            rsrcgen2.set_field( Libgmsec_ruby::U16Field.new("BOGUS-FIELD", 2) )
            rsrcgen2.start()
            check("An exception was expected", false)
        rescue GmsecException => e
            check(e.message, e.message.include?("Message Validation Failed"))
        end
        Libgmsec_ruby::Log::info("Cleanup...")
        Libgmsec_ruby::ResourceGenerator::destroy(rsrcgen2)
    end


    def test_stop()
        Libgmsec_ruby::Log::info("Test stop()")

        config = get_config()
        pubRate = 1

        rsrcgen = Libgmsec_ruby::ResourceGenerator::create( config, pubRate, 1, 10, @@stdFields )

        # To ensure this resource message is unique, we set the COMPONENT field
        rsrcgen.set_field( Libgmsec_ruby::StringField.new("COMPONENT", get_unique_component()) )

        rsrcgen.start()

        check("ResourceGenerator should be running", rsrcgen.is_running())

        rsrcgen.stop()

        check("ResourceGenerator should not be running", rsrcgen.is_running() == false)

        # Allow time for the RSRC-gen thread to stop
        sleep(4)

        conn = Libgmsec_ruby::Connection::create(config)
        conn.connect()
        conn.subscribe("C2MS.*.*.*.*.*.MSG.RSRC." + get_unique_component())

        sleep(4)

        for i in 0..3 do
            rsrcMsg = conn.receive(5000)

            if rsrcMsg != nil
                check("Unexpectedly received a Resource Message", false)
                Libgmsec_ruby::Message::destroy(rsrcMsg)
            end
        end

        conn.disconnect()

        Libgmsec_ruby::Connection::destroy(conn)
        Libgmsec_ruby::ResourceGenerator::destroy(rsrcgen)
    end


    def test_set_field()
        Libgmsec_ruby::Log::info("Test set_field()")

        config = Libgmsec_ruby::Config.new( get_config() )     # make a copy
        pubRate = 1

        # Nominal tests
        rsrcgen = Libgmsec_ruby::ResourceGenerator::create( config, pubRate, 1, 10, @@stdFields )

        # To ensure this resource message is unique, we set the COMPONENT field
        rsrcgen.set_field( Libgmsec_ruby::StringField.new("COMPONENT", get_unique_component()) )

        rsrcgen.start()

        sleep(4)

        Libgmsec_ruby::Log::info("Test set_field() with a Field")
        newPubRate = 2
        rsrcgen.set_field( Libgmsec_ruby::U16Field.new("PUB-RATE", newPubRate) )

        verify_resource_message(config, newPubRate)

        rsrcgen.stop()

        # Allow time for the RSRC-gen thread to stop
        sleep(4)

        Libgmsec_ruby::Log::info("Cleanup...")
        Libgmsec_ruby::ResourceGenerator::destroy(rsrcgen)
    end


    def verify_resource_message(config, expectedPubRate)
        begin
            conn = Libgmsec_ruby::Connection::create(config)
            conn.connect()
            conn.subscribe("C2MS.*.*.*.*.*.MSG.RSRC." + get_unique_component() + ".+")

            sleep(4)

            rsrcMsg = conn.receive(5000)

            if rsrcMsg != nil
                check("Unexpected MESSAGE-TYPE", rsrcMsg.get_string_value("MESSAGE-TYPE") == "MSG")
                check("Unexpected MESSAGE-SUBTYPE", rsrcMsg.get_string_value("MESSAGE-SUBTYPE") == "RSRC")
                check("Unexpected PUB-RATE", rsrcMsg.get_integer_value("PUB-RATE") == expectedPubRate)

                Libgmsec_ruby::Message::destroy(rsrcMsg)
            else
                check("Failed to received Resource Message", false)
            end

            conn.disconnect()
            Libgmsec_ruby::Connection::destroy(conn)
        rescue GmsecException => e
            require("verify_resource_message: " + e.message, false);
        end
    end
end


begin
    test = Test_ResourceGenerator.new( ARGV )
    test.run
end

File: tests/nodejs/src/T016_ResourceGenerator.js
#!/usr/bin/env node

const TestCase = require('./TestCase.js');
const gmsec = require('../../../bin/gmsec_nodejs.node');


class T016_ResourceGenerator extends TestCase
{
	constructor()
	{
		super();
	}

	testCase()
	{
		this.testConstructor_1();
		this.testConstructor_2();
		this.testStart();
		this.testStop();
		this.testSetField();
		this.testCreateResourceMessage();
	}

	testConstructor_1()
	{
		gmsec.Log.info("Test constructor w/ no fields");

		var config1 = new gmsec.Config(this.getConfig());
		var config2 = new gmsec.Config(this.getConfig());
		var config3 = new gmsec.Config("mw-id=bogus", gmsec.DataType_KEY_VALUE_DATA);

		config2.addValue("gmsec-msg-content-validate", "true");

		// Nominal test
		try {
			var rsrcgen = new gmsec.ResourceGenerator(config1, 5, 1, 10);
			this.check("ResourceGenerator should not be running", rsrcgen.isRunning() == false);
		}
		catch (e) {
			this.require(e.message, false);
		}

		// Off-nominal tests
		// TODO: Not able to test because SWIG-generated code does not propagate
		// GMSEC API exceptions on overloaded functions.
		/*
		try {
			// Missing standard fields to make RSRC message compliant
			var rsrcgen = new gmsec.ResourceGenerator(config2, 5, 1, 10);
			this.check("Expected an exception", false);
		}
		catch (e) {
			this.check(e.message, e.message.includes("Message Validation Failed"));
		}

		try {
			// Bogus middleware
			var rsrcgen = new gmsec.ResourceGenerator(config3, 5, 1, 10);
			this.check("Expected an exception", false);
		}
		catch (e) {
			this.check(e.message, e.message.includes("Unable to load"));
		}
		*/
	}

	testConstructor_2()
	{
		gmsec.Log.info("Test constructor w/ fields");

		var config1 = new gmsec.Config(this.getConfig());
		var config2 = new gmsec.Config(this.getConfig());
		var config3 = new gmsec.Config("mw-id=bogus", gmsec.DataType_KEY_VALUE_DATA);

		config2.addValue("gmsec-msg-content-validate", "true");

		var emptyFieldArray = new gmsec.FieldArray();
		var standardFields  = this.getStandardFields();

		try {
			var rsrcgen1 = new gmsec.ResourceGenerator(config1, 5, 1, 10, emptyFieldArray);
			var rsrcgen2 = new gmsec.ResourceGenerator(config1, 5, 1, 10, standardFields);

			this.check("RSRC Gen 1 should not be running", rsrcgen1.isRunning() == false);
			this.check("RSRC Gen 2 should not be running", rsrcgen2.isRunning() == false);
		}
		catch (e) {
			this.require(e.message, false);
		}

		// Off-nominal tests
		// TODO: Not able to test because SWIG-generated code does not propagate
		// GMSEC API exceptions on overloaded functions.
		/*
		try {
			// Missing standard fields to make RSRC message compliant
			var rsrcgen = new gmsec.ResourceGenerator(config2, 5, 1, 10, emptyFieldArray);
			this.check("Expected an exception", false);
		}
		catch (e) {
			this.check(e.message, e.message.includes("Message Validation Failed"));
		}

		try {
			// Bogus middleware
			var rsrcgen = new gmsec.ResourceGenerator(config3, 5, 1, 10, emptyFieldArray);
			this.check("Expected an exception", false);
		}
		catch (e) {
			this.check(e.message, e.message.includes("Unable to load"));
		}
		*/
	}

	testStart()
	{
		gmsec.Log.info("Test start()");

		try {
			var pubRate = 1;
			var rsrcgen = new gmsec.ResourceGenerator(this.getConfig(), pubRate, 1, 10, this.getStandardFields());

			// To ensure the RSRC generator encapsulated message is unique, we set the COMPONENT field
			rsrcgen.setField(new gmsec.StringField("COMPONENT", this.getUniqueComponent()));

			// Start the RSRC generator
			rsrcgen.start();
			this.check("RSRC Gen should be running", rsrcgen.isRunning());

			this.verifyResourceMessage(this.getConfig(), pubRate);

			// Stop the RSRC generator
			rsrcgen.stop();
		}
		catch (e) {
			this.require(e.message, false);
		}
	}

	testStop()
	{
		gmsec.Log.info("Test stop()");

		try {
			var pubRate = 1;
			var rsrcgen = new gmsec.ResourceGenerator(this.getConfig(), pubRate, 1, 10, this.getStandardFields());

			// To ensure the RSRC generator encapsulated message is unique, we set the COMPONENT field
			rsrcgen.setField(new gmsec.StringField("COMPONENT", this.getUniqueComponent()));

			// Start the RSRC generator
			rsrcgen.start();
			this.check("RSRC Gen should be running", rsrcgen.isRunning());

			// Delay before stopping the RSRC generator
			gmsec.TimeUtil.millisleep(2000);

			// Stop the RSRC generator
			rsrcgen.stop();
			this.check("RSRC Gen should be running", rsrcgen.isRunning() == false);

			// Allow time for the RSRC (thread) to truly stop
			gmsec.TimeUtil.millisleep(2000);

			// Verify no RSRC messages are being published
			try {
				var conn = gmsec.Connection.create(this.getConfig());
				conn.connect();
				conn.setupSubscription("C2MS.*.*.*.*.*.MSG.RSRC." + this.getUniqueComponent() + ".+");

				for (var i = 0; i < 3; i++) {
					var hbmsg = conn.receive(5000);

					if (hbmsg != null) {
						this.check("Received unexpected RSRC message", false);
					}
				}

				conn.disconnect();

				gmsec.Connection.destroy(conn);
			}
			catch (e) {
				this.require(e.message, false);
			}
		}
		catch (e) {
			this.require(e.message, false);
		}
	}

	testSetField()
	{
		gmsec.Log.info("Test setField()");

		// Nominal tests
		try {
			var pubRate = 1;
			var rsrcgen = new gmsec.ResourceGenerator(this.getConfig(), pubRate, 1, 10, this.getStandardFields());

			// To ensure the RSRC generator encapsulated message is unique, we set the COMPONENT field
			rsrcgen.setField(new gmsec.StringField("COMPONENT", this.getUniqueComponent()));

			rsrcgen.start();

			gmsec.TimeUtil.millisleep(2000);

			gmsec.Log.info("Test setField() with a Field object");
			var newPubRate = 2;
			rsrcgen.setField(new gmsec.U16Field("PUB-RATE", newPubRate));
			this.verifyResourceMessage(this.getConfig(), newPubRate);

			rsrcgen.stop();
		}
		catch (e) {
			this.require(e.message, false);
		}

		// Off-nominal tests
		// TODO: Not able to test because SWIG-generated code does not propagate
		// GMSEC API exceptions on overloaded functions.
		/*
		var config = new gmsec.Config(this.getConfig());
		config.addValue("gmsec-msg-content-validate", "true");

		try {
			// Add bogus field using a Field object
			var rsrcgen = new gmsec.ResourceGenerator(config, 1, 1, 10, this.getStandardFields());
			rsrcgen.setField(new gmsec.U16Field("BOGUS-FIELD", 2));
		}
		catch (e) {
			this.check(e.message, e.message.includes("Message Validation Failed"));
		}
		*/
	}

	testCreateResourceMessage()
	{
		gmsec.Log.info("Test createResourceMessage()");

		var factory = new gmsec.MessageFactory(this.getConfig());
		var samInt  = 1;
		var avgInt  = 10;

		this.setStandardFields(factory);

		var rsrcMsg = gmsec.ResourceGenerator.createResourceMessage(factory, samInt, avgInt);
		this.check("Unexpected MESSAGE-TYPE", rsrcMsg.getStringValue("MESSAGE-TYPE") === "MSG");
		this.check("Unexpected MESSAGE-SUBTYPE", rsrcMsg.getStringValue("MESSAGE-SUBTYPE") === "RSRC");
	}

	verifyResourceMessage(config, pubRate)
	{
		try {
			var conn = gmsec.Connection.create(config);
			conn.connect();
			conn.setupSubscription("C2MS.*.*.*.*.*.MSG.RSRC." + this.getUniqueComponent() + ".+");

			var t1 = 0;
			var t2 = 0;

			for (var i = 0; i < 8; i++) {
				// Start time t1 is at the time start to receive data
				t1 = gmsec.TimeUtil.getCurrentTime_s();
				var hbmsg = conn.receive(5000);

				// The end time t2 has to be measured immediately right after it completes receiving data.
				t2 = gmsec.TimeUtil.getCurrentTime_s();

				// Ignore the first few incoming messages (if any)
				if (i < 3) continue;

				if (hbmsg != null) {
					//if (t1 == 0) {
					//	t1 = gmsec.TimeUtil.getCurrentTime_s();
					//}
					//else {
					//	t2 = gmsec.TimeUtil.getCurrentTime_s();

						var delta = t2 - t1;
						roundUpPubRateToSecond = Math.floor(delta);
						if (delta < pubRate) {
							//delta = ((t2 - t1) * 10.0 + 0.5) / 10.0;
							roundUpPubRateToSecond = Math.floor(delta + 0.5);    // round up to second
						}

						gmsec.Log.info("Expected Rate: " + pubRate + ", delta is: " + delta);

						//this.check("Unexpected publish rate", Math.floor(delta) == pubRate);
						this.check("Unexpected publish rate", roundUpPubRateToSecond == pubRate);

					//	t1 = t2;
					//}
				}
				else {
					this.check("Failed to receive RSRC message", false);
				}
			}

			conn.disconnect();

			gmsec.Connection.destroy(conn);
		}
		catch (e) {
			this.require(e.message, false);
		}
	}
}

test = new T016_ResourceGenerator();
test.setDescription("Test ResourceGenerator");
test.run(process.argv);
